ca/crl.go

```diff
@@ -0,0 +1,221 @@
+package ca
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"strings"
+	"time"
+
+	capb "github.com/letsencrypt/boulder/ca/proto"
+	"github.com/letsencrypt/boulder/core"
+	corepb "github.com/letsencrypt/boulder/core/proto"
+	"github.com/letsencrypt/boulder/issuance"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+type crlImpl struct {
+	capb.UnimplementedCRLGeneratorServer
+	issuers   map[issuance.IssuerNameID]*issuance.Issuer
+	lifetime  time.Duration
+	maxLogLen int
+	log       blog.Logger
+}
+
+func NewCRLImpl(issuers []*issuance.Issuer, lifetime time.Duration, maxLogLen int, logger blog.Logger) (*crlImpl, error) {
+	issuersByNameID := make(map[issuance.IssuerNameID]*issuance.Issuer, len(issuers))
+	for _, issuer := range issuers {
+		issuersByNameID[issuer.Cert.NameID()] = issuer
+	}
+
+	if lifetime == 0 {
+		logger.Warningf("got zero for crl lifetime; setting to default 9 days")
+		lifetime = 9 * 24 * time.Hour
+	} else if lifetime >= 10*24*time.Hour {
+		return nil, fmt.Errorf("crl lifetime cannot be more than 10 days, got: %s", lifetime)
+	} else if lifetime <= 0*time.Hour {
+		return nil, fmt.Errorf("crl lifetime must be positive, got: %s", lifetime)
```

Comment:
To make these messages more useful to an operator you could provide, in quotes, the (implicit) name of this field in the JSON config.

Similarly, quoting the parsed input can be helpful as well.

Reply:
Unfortunately the code here inside the constructor has no knowledge of the config keys used to set these arguments -- I think "crl lifetime" is close enough to "crlLifetime" to be figured out?

I've quoted the values.

Reply:
Works for me!